decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,460 @@
+
+-   **Status:** Proposed
+-   **Impact:** High
+-   **Driver:** @foundrytom
+-   **Approver:** @foundrytom @feltech
+-   **Contributors:** @foundrytom @feltech @meshula
+-   **Outcome:** OpenAssetIO will use `C++` as the core language for
+    efficient implementation, with critical business logic exposed
+    through portable `C` API.
+
+## Background
+
+OpenAssetIO strives to provide a universal standard for communications
+between a process that consumes or produces data, and any external
+systems that manage that data.
+
+To be successful in any meaningful way, it needs to be portable to
+which ever programming language(s) are meaningful to its target
+audience.
+
+At the time of writing, the majority of expected integrations are within
+`C++` and/or `Python` environments. Other languages such as `Rust` or
+`go` are gaining traction. Broader reaching applications of OpenAssetIO
+introduce `JavaScript`, `Swift` and others.
+
+The library must be designed to simplify integration into current
+_and_ future programming environments, without the need to re-implement
+critical business logic.
+
+## Relevant data
+
+### Required language support
+
+`C++`, `Python`
+
+### Projected language support
+
+`C`, `JavaScript`, `Java`, `Swift`, `Rust`, `go`
+
+### Bindings
+
+Research has shown that the common denominator when addressing the above
+listed languages is a `C` interface. A standard `C` API surface, with
+clear and manageable ownership semantics would facilitate support in the
+broadest range of future programming environments.
+
+Ideally, any bindings would also allow idiomatic programming within
+the target language, rather than forcing specific design patterns from
+any internals of OpenAssetIO - such as synchronous programming, RIAA,
+etc.
+
+### Performance
+
+A core requirement raised by numerous parties is that the "happy path"
+should be as performant as possible. Thus, it is required that any
+mechanism introduced to broaden compatibility doesn't compromise the
+(current) majority deployment scenarios.
+
+Of note, a `C++` host talking to a `C++` manager has been the most
+frequently requested high-performance pathway.
+
+### Development lifecycle
+
+The overhead of maintaining dynamically linked `C++` libraries within a
+plugin-based architecture is often fraught with constantly chasing ABI
+changes and/or conflicts between mutual dependencies.
+
+Foundry has prior art in using `C` function pointer suites as a compiler
+isolation technique that can help de-couple the plugin host and the
+specific of any given plugin.
+
+In the constantly evolving landscape defining which dependency versions
+are in use within which versions of common DCC tools, avoiding the need
+to matrix-compile and juggle plugin search paths per-host would be
+massively beneficial.
+
+## Options considered
+
+### Option 1
+
+`C++` and `Python` support only.
+
+**Estimated Cost**: Small
+
+#### Pros
+
+-   Simplifies development.
+
+#### Cons
```

Comment:
[consider] No C++ ABI isolation, so subject to compiler and library ABI breaks between the host, OpenAssetIO and manager plugins.

Reply:
Added [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/ff11aeee4d2e53abf1b3ba5a2e336b354952bbf6).

Reply:
A random thought... not sure if this is useful for you guys, but the rust wg wrote a tool (cppmm) which generates a c api for a c++ library. It's what we're using to make rust bindings.

Reply:
An example of usd stage in c. It's fairly easy to use and the whole project is in c++.

https://github.com/vfx-rs/usd-bind/blob/main/usd-c/include.in/pxr/usd/usd/usd_stage.h

Reply:
Thanks @luke-titley looks really promising, we'll def factor this in when we start on the `C` exposed code for-real.